src/HappyConnOpener.cc

```diff
@@ -630,8 +631,15 @@ HappyConnOpener::handleConnOpenerAnswer(Attempt &attempt, const CommConnectCbPar
     }
 
     debugs(17, 8, what << " failed: " << params.conn);
-    if (const auto peer = params.conn->getPeer())
-        peerConnectFailed(peer);
+    if (const auto peer = params.conn->getPeer()) {
+        std::unique_ptr<ACLFilledChecklist> ch;
+        if (Config.accessList.cachePeerFault) {
+            ch.reset(new ACLFilledChecklist(Config.accessList.cachePeerFault, cause.getRaw(), nullptr));
+            ch->al = ale;
+            ch->syncAle(cause.getRaw(), nullptr);
+        }
+        peer->peerConnectFailed(ch.get());
```

Comment:
It feels like we are doing the same "do we need to check the ACLs?" work twice, once here and once inside peerConnectFailed(). And we are introducing heap allocations/pointer complications for an essentially local/temporary checklist object.

I see only one way to improve it -- lambdas. Please consider doing something along these lines instead:

caller:

```C++
// TODO: Use auto as the lambda parameter type after upgrading to C++14
peer->noteFailure([](ACLFilledChecklist &checklist) {
    checklist.setRequest(cause.getRaw());
    checklist.al = ale;
    checklist.syncAle(cause.getRaw(), nullptr);
});
```

The above TODO comment can go in just one caller even though it applies to all of them.

CachePeer.h:

```C++
public:
    /// reacts to a cache_peer-related failure
    /// \param filler a function to configure an ACLFilledChecklist if that becomes necessary
    void noteFailure(... filler);

private:
    void countFailure();
```

I do not think we need templates here, but I did not spend the time to provide the exact parameter type -- I just used a `...` placeholder. Please keep the actual filler function type as simple and as constant as possible.


CachePeer.cc:

```C++
void
CachePeer::noteFailure(... filler)
{
    if (const auto acls = Config.accessList.cachePeerFault) {
        ACLFilledChecklist checklist(acls, nullptr, nullptr));
        filler(checklist);
        if (!checklist.fastCheck().allowed())
            return; // this failure is not our fault
    }
    // else take the blame for all failures

    countFailure();
}
```